pkg/maintainer/spv/spv.go

```diff
@@ -59,8 +63,271 @@ func (sm *spvMaintainer) startControlLoop(ctx context.Context) {
 func (sm *spvMaintainer) maintainSpv(ctx context.Context) error {
 	logger.Infof("Maintaining SPV proof...")
 
-	// TODO: Implement. For now, just wait.
-	<-ctx.Done()
+	for {
+		if err := sm.proveDepositSweepTransactions(); err != nil {
+			return fmt.Errorf(
+				"error while proving deposit sweep transactions: [%v]",
+				err,
+			)
+		}
+
+		// TODO: Add proving of other type of SPV transactions: redemption
+		// transactions, moving funds transaction, etc.
+
+		select {
+		case <-time.After(sm.config.IdleBackOffTime):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (sm *spvMaintainer) proveDepositSweepTransactions() error {
+	depositSweepTransactions, err := sm.getUnprovenDepositSweepTransactions()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get unproven deposit sweep transactions: [%v]",
+			err,
+		)
+	}
+
+	// TODO: Consider handling a situation in which the block headers in the
+	//       proof span multiple Bitcoin difficulty epochs.
+	requiredConfirmations, err := sm.chain.TxProofDifficultyFactor()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get transaction proof difficulty factor: [%v]",
+			err,
+		)
+	}
+
+	for _, transaction := range depositSweepTransactions {
+		_, proof, err := bitcoin.AssembleSpvProof(
+			transaction.Hash(),
+			uint(requiredConfirmations.Uint64()),
+			sm.btcChain,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to assemble SPV proof: [%v]", err)
+		}
+
+		mainUTXO, vault, err := parseTransactionInputs(
+			sm.btcChain,
+			sm.chain,
+			transaction,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"error while parsing transaction inputs: [%v]",
+				err,
+			)
+		}
+
+		if err := sm.chain.SubmitDepositSweepProofWithReimbursement(
+			transaction,
+			proof,
+			mainUTXO,
+			vault,
+		); err != nil {
+			return fmt.Errorf(
+				"failed to submit deposit sweep proof with reimbursement: [%v]",
+				err,
+			)
+		}
+	}
 
 	return nil
 }
+
+func (sm *spvMaintainer) getUnprovenDepositSweepTransactions() (
+	[]*bitcoin.Transaction,
+	error,
+) {
+	// TODO: Limit how far in the past we are looking for the events.
+	//       Possibly store latest checked height in memory or file.
+	depositSweepTransactionProposals, err :=
+		sm.chain.PastDepositSweepProposalSubmittedEvents(nil)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to get past deposit sweep proposal submitted events: [%v]",
+			err,
+		)
+	}
+
+	// There will often be multiple events emitted for a single wallet. Prepare
+	// a list of unique wallet public key hashes.
+	walletPublicKeyHashes := uniqueWalletPublicKeyHashes(
+		depositSweepTransactionProposals,
+	)
+
+	unprovenDepositSweepTransactions := []*bitcoin.Transaction{}
+
+	for _, walletPublicKeyHash := range walletPublicKeyHashes {
+		// TODO: Should we check the wallet's state before attempting to submit
+		//       the deposit sweep proof?
```

Comment:
Hmm... why? I can't think of a scenario when we should wait with providing SPV proof but of course, I could miss something. In my opinion, it should happen as soon as possible - if the wallet was late with signing, providing SPV proof right away will allow the next action to be performed.

Reply:
We could filter out wallets that are not `Live` or `MovingFunds` for performance reasons.
If the wallet is `Closed`, we could skip it immediately.

Reply:
That's a good point, I thought this is about the wallet lock state in the `WalletCoordinator` contract. If this is easy, I think we should do it. This is not an issue now when all wallets are Live but we'll get to the point when this check will matter.

Reply:
Added a check for wallet state in 8abdeecdac4f3cc85fce44b60968bab6d68b7d94.